### DISCUSSION TITLE
Installation Of Sorotoki

### DISCUSSION THREAD

**Discussion by gugulothus41** [BJCaasenbrood/SorotokiCode#12] [Q&A] [answered]
Dear Sir,
I curious to learn Soft robotics and its mechanisms, it great herd that Sorokati is one of the best platform.
I am getting installation problem, where can i get these, do i need to add-on in existing Matlab.
Can you briefly explain to installation.

**Comment by BJCaasenbrood** [comment]
Dear @gugulothus41,

Happy to hear your are taking an interest in SOROTOKI. I wouldn't say SOROTOKI is the best platform, I think [SoroSim](https://github.com/Ikhlas-Ben-Hmida/SoRoSim), [PyElastica](https://github.com/GazzolaLab/PyElastica), and [SOFA](https://github.com/sofa-framework/sofa) easily outperform my toolkit, but I take the compliment haha :wink:.

There is two ways of installing SOROTOKI, via `mpm` (preferred method) and unzipping the toolkit, and loading the path using `addpath`. The first method is a lot easier, and the second method I only recommend to more experienced users of matlab.

For your information, there is an installation guide provided here: https://bjcaasenbrood.github.io/SorotokiCode/install/

To summarize, you need to following:
1) Install the [Matlab Package Manager](https://nl.mathworks.com/matlabcentral/fileexchange/54548-mpm?s_tid=srchtitle) (MPM) via the Matlab Add-On Mananger ![](https://nl.mathworks.com/help/matlab/matlab_env/add-ons_24x24.png). You can also find the package here at the MATLAB file exchange: https://nl.mathworks.com/matlabcentral/fileexchange/54548-mobeets-mpm.
2) Install SOROTOKI via mpm
```matlab
mpm install sorotokicode 
```
3) Navigate to where SOROTOKI is installed, typically under `./MATLAB Add-Ons/Collections/mpm/mobeets-mpm-c626429/mpm-packages/SorotokiCode`
4) Call the installer. The installer will also check for any missing packages, download these automatically, and adds them to the folder `mpm-packages`. Note though, you require some MATLAB toolboxes, like the Partial Differential Equation (PDE) and Coder toolbox.
```
sorotoki install
```
5) Call the `c++` mex builder:
```
sorotoki build
```
6) (Optional) Check if the toolkit is installed correctly using the testsuite:
```matlab
sorotoki test
```

**Reply by gugulothus41** [reply]
![Sorotoki_facing_Problem](https://github.com/BJCaasenbrood/SorotokiCode/assets/158456738/1f5f19c1-512d-4a57-8b0c-fbdf37e49cc5)
![Sorotoki_facing_Problem](https://github.com/BJCaasenbrood/SorotokiCode/assets/158456738/999f73f2-0d8d-4171-a690-aad54dd93ee9)
.....Please check it is showing error using mpm & i installed necesary tools, Thank you

**Reply by BJCaasenbrood** [reply]
A couple of things. 

First, you are currently in the `sorotokicode/example` folder, whereas `@sorotoki` is located in `sorotokicode` folder. This means that MATLAB cannot find the code associated to the sorotoki installer. Changing the folder to `sorotokicode` should fix this issue. 

Second, regarding the `mpm` issue, this is a common problem also mentioned in the documentation: https://bjcaasenbrood.github.io/SorotokiCode/install/ The issue is that MATLAB has its own built-in function called `mpm`, the two are in conflict. We've addressed this issue with the developer of mpm (@mobeets), and posted an issue on https://github.com/mobeets/mpm/issues/82 Currently, there is not fix for this issue except for changing the name of mpm.

A fix that worked for me is making a file called `startup.m` in the folder `./Documents/MATLAB/` folder. This file is called during the startup of MATLAB. Here is what my file looks like:

```matlab
warning off

% change USER to your username!
USER = 'brandon';
if ispc
    addpath(genpath([`C:\Users\`,USER,'AppData\',...
            'Roaming\Mathworks\MATLAB Add-Ons\Collections\mpm\mobeets-mpm-c626429']));
elseif isunix
    addpath(genpath(['/home/',USER,'/MATLAB Add-Ons/Collections/mpm/mobeets-mpm-c626429']));
end

mpm init
clc; warning on
```

You should change it to `USER = prgii`. Hopefully this helps. To check if sorotoki is installed correctly, call `sorotoki -t` to test it. Also make sure to build the sorotoki toolkit using `sorotoki -b --approve` (the command `--approve` will automatically reply `y` to all requests).

**Reply by gugulothus41** [reply]
![FinalIssueSoro](https://github.com/BJCaasenbrood/SorotokiCode/assets/158456738/76f2c6d5-9efc-4960-8f77-ae07c489eb61)
I think there change in directory problem for mpm and add-on, is these correct, how can i do.

**Reply by BJCaasenbrood** [reply]
The error you see is still Matlab's built-in mpm, and not the correct mpm installer. How does your 'startup.m' file look like? 

Alternatively, you can add the folder 'mobeets_mpm' to Matlab's path. Click the 'set path' button under preferences. The problem is that mpm.m in the mobeets_mpm folder is not found by Matlab

**Reply by gugulothus41** [reply]
So, far i don't  know which path i added, but i achieved till here, when i am performing backbone_test example..its giving following error.
![SoRoToki_Expl01](https://github.com/BJCaasenbrood/SorotokiCode/assets/158456738/29ebf262-f3fa-49bc-b6a7-1a2871e1fe97)
How can i execute examples Now?

**Reply by BJCaasenbrood** [reply]
Did you call `sorotoki build`? Otherwise the .mex files are not generated...

**Reply by BJCaasenbrood** [reply]
I've updated the README.md to make this installation process a bit more clear. Sorry for all the confusion :)

**Reply by BJCaasenbrood** [reply, marked as answer]
@gugulothus41 I've made a video explaining the installation of Sorotoki. You can check it here: https://www.youtube.com/embed/uwxvwHAUrtQ?si=1s2J80y5O_cwC76s

**Reply by Ouwen1979** [reply]
com./sorotokimodel/mex/forwardkin build: ..... Failed ✗

**Reply by Ouwen1979** [reply]
./sorotokimodel/mex/forwardkin build: ...Index expression out of bounds. Attempted to access element 4. The valid range is 1-3.

**Reply by BJCaasenbrood** [reply]
@Ouwen1979 Where is your SorotokiCode folder located? It has to be under `Documents/MATLAB/mpi-packages/sorotokicode`

**Comment by BJCaasenbrood** [comment]
Also, I am currently in the process of making a number of Youtube videos explaining how to install and use the toolkit. I'll keep you informed if these will be released.

**Reply by gugulothus41** [reply]
Great to hear sir, Basically i am working in Lattice meta-material, where effective properties are estimated only by rigid materials, i am thinking to extent the same idea by Soft materials, So i need brief insight in-terms of modelling and 3D printing with a particular variable density and Young's modulus, Could i achieve these and analyze. I hope this points are also taken into consideration while making you-tube videos. Thank you.

**Reply by BJCaasenbrood** [reply]
In general, Sorotoki is suited for these problems, yes. However, note that metamaterials may have sharp transitions in their dynamics, for example, buckling. Since Sorotoki explores an adaptive Newton-Rhapson scheme, these sharp transitions may lead to instability when solving. Also, lattice materials may require a high resolution mesh, using a MATLAB solver for such requirements is really recommended... For many scenarios, Sorotoki is better suited for 2D FEM problems; 3D problems are possible but may take hours to simulate...

**Comment by Ouwen1979** [comment]
欧文|小欧文
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "BJCaasenbrood/SorotokiCode"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年5月15日(星期三) 下午4:44
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [BJCaasenbrood/SorotokiCode] Installation Of Sorotoki (Discussion #12)





 
@Ouwen1979 Where is your SorotokiCode folder located? It has to be under Documents/MATLAB/mpi-packages/sorotokicode

**Reply by BJCaasenbrood** [reply]
@Ouwen1979 I don't understand...

**Reply by Ouwen1979** [reply]
it is ok.Thank you

**Comment by Ouwen1979** [comment]
It is ok,Thank you&nbsp;


欧文|小欧文
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Brandon ***@***.***&gt;; 
发送时间: 2024年5月15日(星期三) 下午5:27
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [BJCaasenbrood/SorotokiCode] Installation Of Sorotoki (Discussion #12)














 
@Ouwen1979 I don't understand...

**Comment by Ouwen1979** [comment]
when i am performing backbone_test example..its giving following error.函数或变量 'expmapSE3_inline' 无法识别。

![1718266484930](https://github.com/BJCaasenbrood/SorotokiCode/assets/132358808/dc9ceda6-7790-446f-845b-d8b7b0d2f6bd)
